### PR TITLE
Removal of AArch64 macOS from early access in 0.35

### PR DESCRIPTION
**Pull request by Sreekala-Gopakumar**
https://github.com/eclipse-openj9/openj9-docs/issues/992

Updated the related topics.

Signed-off-by: Sreekala Gopakumar <sreekala.gopakumar@ibm.com>

### CODE REVIEW

**Comment by Sreekala-Gopakumar** [comment]
Jenkins doc stage

**Comment by Sreekala-Gopakumar** [comment]
@pshipton - I have added the operating system and compiler level information of macOS AArch64. Please check and confirm. Thanks!

**Comment by pshipton** [comment]
- Aarch64 macOS only supports macOS 11 and later. We should add macOS 11 and 12 to the table for x86 as well, and set them as explicitly supported
- we should mention that -Xcompressedrefs is not supported, fyi the note in https://github.com/eclipse-openj9/openj9/blob/master/doc/release-notes/0.35/0.35.md
- some default settings may be different from x86 macOS. I see -Xmso for Java 17 is 256KB, and not 512KB like x86. The default page size is shown as 16K, not 4K.
- we should check other settings as well, such as the signal list. @knn-k are the signals the same on macOS x86 and aarch64? Do you see anything else that needs to be updated?

Java 11 default settings
```
  -Xmca32K        RAM class segment increment
  -Xmco128K       ROM class segment increment
  -Xmcrs0K        compressed references metadata initial size
  -Xmns2M         initial new space size
  -Xmnx1G         maximum new space size
  -Xms8M          initial memory size
  -Xmos6M         initial old space size
  -Xmox4094M      maximum old space size
  -Xmx4G          memory maximum
  -Xmr4K          remembered set size
  -Xlp:objectheap:pagesize=16K	 large page size
                  available large page sizes:
                  16K
  -Xlp:codecache:pagesize=16K	 large page size for JIT code cache
                  available large page sizes for JIT code cache:
                  16K
  -Xmso256K       operating system thread stack size
  -Xiss2K         java thread stack initial size
  -Xssi16K        java thread stack increment
  -Xss1M          java thread stack maximum size
  -XX:SharedCacheHardLimit=300M shared class cache size
  -Xscmx64M       shared class cache soft max size
  -Xscdmx24544K   reserved shared class cache space for class debug attributes
  -Xscminaot0K    min reserved shared class cache space for AOT
  -Xscmaxaot282288K max allowed shared class cache space for AOT
  -Xscminjitdata0K min reserved shared class cache space for JIT data
  -Xscmaxjitdata282288K max allowed shared class cache space for JIT data
```

Java 17 default settings
```
  -Xmca32K        RAM class segment increment
  -Xmco128K       ROM class segment increment
  -Xmcrs0K        compressed references metadata initial size
  -Xmns2M         initial new space size
  -Xmnx1G         maximum new space size
  -Xms8M          initial memory size
  -Xmos6M         initial old space size
  -Xmox4094M      maximum old space size
  -Xmx4G          memory maximum
  -Xmr4K          remembered set size
  -Xlp:objectheap:pagesize=16K	 large page size
                  available large page sizes:
                  16K
  -Xlp:codecache:pagesize=16K	 large page size for JIT code cache
                  available large page sizes for JIT code cache:
                  16K
  -Xmso256K       operating system thread stack size
  -Xiss2K         java thread stack initial size
  -Xssi16K        java thread stack increment
  -Xss1M          java thread stack maximum size
  -XX:SharedCacheHardLimit=300M shared class cache size
  -Xscmx64M       shared class cache soft max size
  -Xscdmx24544K   reserved shared class cache space for class debug attributes
  -Xscminaot0K    min reserved shared class cache space for AOT
  -Xscmaxaot282288K max allowed shared class cache space for AOT
  -Xscminjitdata0K min reserved shared class cache space for JIT data
  -Xscmaxjitdata282288K max allowed shared class cache space for JIT data
`

**Comment by pshipton** [comment]
We should refer to it as "Apple Silicon macOS" and not AArch64

File: docs/version0.35.md [hunk 1]
```diff
@@ -42,6 +42,8 @@ Eclipse OpenJ9&trade; release 0.35.0 supports OpenJDK 8, 11, and 17.
 
 OpenJ9 Windows&reg; builds for OpenJDK 11 are now compiled with Microsoft&reg; Visual Studio 2019. The Visual Studio redistributable files included with the build are updated to match.
 
+AArch64 macOS&reg; is now a fully supported, production-ready target for OpenJDK 17.
```
Comment (pshipton) [inline docs/version0.35.md:45]: Also OpenJDK 11 for this release, and newer releases like 19/20 in the future.

File: docs/openj9_support.md [hunk 2]
```diff
@@ -132,9 +132,9 @@ OpenJDK 11 binaries are expected to function on the minimum operating system lev
...
 | Windows Server 2019                       | :fontawesome-solid-check:{: .yes aria-hidden="true"}<span class="sr-only">yes</span> |
 
-| macOS                                     | x64                                                                                  |
-|-------------------------------------------|--------------------------------------------------------------------------------------|
-| OS X 10.11+                               | :fontawesome-solid-check:{: .yes aria-hidden="true"}<span class="sr-only">yes</span> |
+| macOS         | x64                                                                                  | AArch64                                                                                                |
+|---------------|--------------------------------------------------------------------------------------|---------------------------------------------------------------------------------|
+| OS X 10.11+   | :fontawesome-solid-check:{: .yes aria-hidden="true"}<span class="sr-only">yes</span> | :fontawesome-solid-check:{: .yes aria-hidden="true"}<span class="sr-only">yes</span> |
```
Comment (knn-k) [inline docs/openj9_support.md:137]: The AArch64 build does not support OS X 10.15.7 and older.

**Comment by knn-k** [comment]
> Do you see anything else that needs to be updated?

Please let me check it tomorrow.

**Comment by knn-k** [comment]
- I have made no particular changes for signal handling, and I think the AArch64 macOS port handles signals in the same way as the existing Intel macOS port.  Is there any way to get a list of signals handled by the OpenJ9 VM?
- I have made no particular changes for the native stack size (`-Xmso`) during the port, either.
- The default memory page size is 16KB on AArch64 macOS as the command `sysctl -a vm.pagesize` answers the value.  The OpenJ9 VM just uses the size.

**Comment by pshipton** [comment]
The list of signals for x86 macOS are here https://eclipse-openj9.github.io/openj9-docs/openj9_signals/

**Comment by knn-k** [comment]
Ah, yes, I know the list there.
My question was: Is there any command-line option or any other way to generate a list of handled signals?

File: docs/openj9_support.md [hunk 3]
```diff
@@ -172,9 +172,9 @@ OpenJDK 17 binaries are expected to function on the minimum operating system lev
...
 | Windows Server 2019                       | :fontawesome-solid-check:{: .yes aria-hidden="true"}<span class="sr-only">yes</span> |
 
-| macOS                                     | x64                                                                                  |
-|-------------------------------------------|--------------------------------------------------------------------------------------|
-| OS X 10.14+                               | :fontawesome-solid-check:{: .yes aria-hidden="true"}<span class="sr-only">yes</span> |
+| macOS         | x64                                                                                  | AArch64                                                                               |
+|---------------|--------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------|
+| OS X 10.14+   | :fontawesome-solid-check:{: .yes aria-hidden="true"}<span class="sr-only">yes</span> | :fontawesome-solid-check:{: .yes aria-hidden="true"}<span class="sr-only">yes</span> |
```
Comment (knn-k) [inline docs/openj9_support.md:177]: The table for OpenJDK 17 on macOS should look like this.

```
|--------------|-----|---------------|
| macOS        | x64 | Apple Silicon |
|--------------|-----|---------------|
| macOS 10.14+ | v   | N/A           |
| macOS 11     | v   | v             |
| macOS 12     | v   | v             |
|--------------|-----|---------------|
```

"OS X" is the old name for the versions before macOS Sierra (10.12).

**Comment by Sreekala-Gopakumar** [comment]
> We should refer to it as "Apple Silicon macOS" and not AArch64

Some clarification required:

1. Should all instances of "AArch64" be replaced with "Apple Silicon macOS" in the entire user guide?

2. While reading about this in the Wikipedia, it is written that "Apple silicon chips 'mainly' use the ARM architecture (of which AArch64 is a variety)". So, can we consider "AArch64" same as "Apple silicon"?

3. Should we be saying "Apple Silicon" or "Apple silicon"? On Apple's website, it seems to be used as "Apple silicon", i.e. silicon chips from Apple. "Apple Silicon" doesn't appear in Apple's trademark list.

Thanks!

**Comment by pshipton** [comment]
> can we consider "AArch64" same as "Apple silicon"

It's possible to have AArch64 which is not Apple silicon / macOS. i.e. we support AArch64 Linux. I expect macOS only supports Apple silicon, and not AArch64 in general.

> Should we be saying "Apple Silicon" or "Apple silicon" ...

"Apple silicon" according to your research.

**Comment by pshipton** [comment]
@knn-k 

> Is there any command-line option or any other way to generate a list of handled signals?

No, just by examining the code.

**Comment by knn-k** [comment]
@pshipton 
OK, I looked through the signal handling code, and I see no Intel-specific or AArch64-specific code for macOS.  So I have nothing to add/change the signal list.
I don't have anything else to update the documents for AArch64 macOS, either.

**Comment by doveye** [comment]
As far as I can tell, "Apple silicon" is basically a term that collectively refers to processors made by Apple (they talk about their "custom silicon" on their website and the Wikipedia page about "Apple silicon" covers processors going back over a decade). It seems to be the equivalent of "Intel" or "AMD" (the `Kind` field in System Information->Software->Applications on a mac apparently has the options "Apple Silicon", "Intel", and "Universal"). The actual architecture appears to be AArch64, certainly for the M1 processor (as shown in the tables at https://en.wikipedia.org/wiki/Apple_silicon#Apple_M1). But it's a custom Apple-built processor that uses the AArch64 instruction set on licence from ARM, not a standard AArch64 processor from ARM. So it's different from the AArch64 processors in other computers, such as those that might be running Linux (though apparently this doesn't affect the code much, [according to Apple](https://developer.apple.com/documentation/xcode/writing-arm64-code-for-apple-platforms)).

So where we talk about architecture (eg in the tables in https://www.eclipse.org/openj9/docs/openj9_support/), I think we should use "AArch64", maybe with "Apple silicon" in brackets afterwards if we think it's not clear enough otherwise (or just "Apple" perhaps). And maybe we also need to change the existing instances of "AArch64" that relate to Linux to "AArch64 (Arm)", especially since people are working on getting Linux to run natively on the Apple processors. (Extra confusion: "Arm" is the company name whereas "ARM" is their processor.)

**Comment by pshipton** [comment]
> Does this mean that -Xmso for Java 17 is 512KB like x86 and not 256KB

It's 256KB, as shown in https://github.com/eclipse-openj9/openj9-docs/pull/1010#issuecomment-1277501665 under "Java 17 default settings", `-Xmso256K       operating system thread stack size`.

**Comment by Sreekala-Gopakumar** [comment]
Jenkins doc stage

**Comment by Sreekala-Gopakumar** [comment]
@pshipton @knn-k - Incorporated the suggested changes. Please check and let me know if any more changes are required. Thanks!

File: docs/openj9_support.md (hunk 2 above)
Comment (knn-k) [reply docs/openj9_support.md:143]: Please remove the row for macOS 10.15.7+, and add rows for macOS 11 and 12 instead.
For Java 11 and AArch64, 
- OS X10.11+ has to be no.
- macOS 11 has to be yes.
- macOS 12 has to be yes.

**Comment by Sreekala-Gopakumar** [comment]
Jenkins doc stage

File: docs/openj9_support.md (hunk 2 above)
Comment (Sreekala-Gopakumar) [reply docs/openj9_support.md]: Done.

Comment (knn-k) [reply docs/openj9_support.md]: LGTM.

**Comment by Sreekala-Gopakumar** [comment]
@pshipton - If there are no other review comments, could you please merge the documents? Thanks!

**Comment by pshipton** [comment]
Merging is waiting on final testing to confirm AArch64 Mac is leaving EA status.

**Comment by Sreekala-Gopakumar** [comment]
@pshipton - Please check and merge if OK. Thanks!

**Review by pshipton — APPROVED** [review]

**Pull request merged**